buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -565,10 +565,25 @@ static ByteBuf encodeString0(ByteBufAllocator alloc, boolean enforceHeap, CharBu
         }
     }
 
+    static String readUtf8(AbstractByteBuf buf, int readerIndex, int length) { // Fast-Path for a AbstractByteBuf
+        buf.checkIndex(readerIndex, length);
```

Comment:
Apparently the JIT doesn't want to inline ByteBuf.forEachByte(), and even when I get it to (by checking instanceof AbstractByteBuf), it refuses to inline and stack-allocate UTFProcessor.